src/Symfony/Component/Console/Application.php

```diff
@@ -179,7 +179,9 @@ public function doRun(InputInterface $input, OutputInterface $output)
 
         if (!$name) {
             $name = $this->defaultCommand;
-            $input = new ArrayInput(array('command' => $this->defaultCommand));
+            $inputArgument = $this->definition->getArgument('command');
+            $inputArgument->setMode(InputArgument::OPTIONAL);
```

Comment:
We could get rid of the setter by redefining the argument entirely e.g

```diff
- $inputArgument = $this->definition->getArgument('command');
- $inputArgument->setMode(InputArgument::OPTIONAL); 
- $inputArgument->setDefault($name);
+ $commandArg = $this->definition->getArgument('command');
+ $this->definition->setArguments(array_merge(
+     $this->definition->getArguments(), 
+     array('command' => new InputArgument('command', InputArgument::OPTIONAL, this->definition->getArgument('command')->getDescription(), $name))
+ ));
```

(not tested)